### PR TITLE
Schema based namespaces

### PR DESCRIPTION
**Pull request by rasputino**
# Schema based namespaces
As it was requested in issue #1075 has been added the option of add the schema as a suffix in the namespaces. It can be used with or without the schema based folders option.

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## Notes
The schema name will end with the `Schema` suffix in the namespace to avoid name conflicts with another entities, maybe this suffix should be parameterizable.

# How Has This Been Tested?
- I added a ScaffoldingPostgreSqlMultiSchemaTest project that reverse engineers the AdventureWorks database that has multiple schemas.

### CODE REVIEW

**Comment by ErikEJ** [comment]
Why is the suffix needed?

**Comment by rasputino** [comment]
> Why is the suffix needed?

Because without it, it could interfere with some entity names. For example, in AdventureWorks model if the suffix doesn't exist then thre will be a lot of errors: `'Person' is a namespace but is used like a type`, because there is ambiguity between the Person schema and the Person entity.

File: src/GUI/RevEng.Core.60/Functions/SqlServerFunctionScaffolder.cs [hunk 1]
```diff
@@ -47,6 +47,12 @@ protected override string WriteDbContext(ModuleScaffolderOptions scaffolderOptio
             Sb.AppendLine("using System.Data;");
             Sb.AppendLine("using System.Linq;");
             Sb.AppendLine($"using {scaffolderOptions.ModelNamespace};");
+
+            if (scaffolderOptions.UseSchemaNamespaces)
+            {
+                model.Routines.Select(r => r.Schema).Distinct().ToList().ForEach(schema => Sb.AppendLine($"using {scaffolderOptions.ModelNamespace}.{schema}"));
```
Comment (ErikEJ) [inline src/GUI/RevEng.Core.60/Functions/SqlServerFunctionScaffolder.cs:53]: This only selects the schemas from stored procedures and functions, this seem wrong

File: src/GUI/RevEng.Core.60/Procedures/SqlServerStoredProcedureScaffolder.cs [hunk 2]
```diff
@@ -233,6 +233,11 @@ private static List<string> CreateUsings(ModuleScaffolderOptions scaffolderOptio
                 $"using {scaffolderOptions.ModelNamespace}",
             };
 
+            if (scaffolderOptions.UseSchemaNamespaces)
+            {
+                model.Routines.Select(r => r.Schema).Distinct().ToList().ForEach(schema => usings.Add($"using {scaffolderOptions.ModelNamespace}.{schema}"));
```
Comment (ErikEJ) [inline src/GUI/RevEng.Core.60/Procedures/SqlServerStoredProcedureScaffolder.cs:238]: Only selects routines

File: src/GUI/RevEng.Core.60/ReverseEngineerScaffolder.cs [hunk 3]
```diff
@@ -332,9 +339,57 @@ private static void AppendSchemaFolders(IModel databaseModel, ScaffoldedModel sc
...
+            {
+                var newUsing = regexStartsWithNameSpace.Replace(newNameSpaceLine, usingKeyWord, 1);
+                newUsings.Append(newUsing);
+                newUsings.Append('.');
+                newUsings.Append(schema);
+                newUsings.Append(nameSpaceSuffix);
+                newUsings.AppendLine(";");
+            }
+
+            newNameSpaceLine = newNameSpaceLine + $".{entityTypeSchemaWithSuffix}";
```
Comment (ErikEJ) [inline src/GUI/RevEng.Core.60/ReverseEngineerScaffolder.cs:373]: Would be great with some sorting of the namespaces

File: src/GUI/RevEng.Core.Abstractions/ModuleScaffolderOptions.cs [hunk 4]
```diff
@@ -11,5 +11,7 @@ public class ModuleScaffolderOptions
         public virtual bool NullableReferences { get; set; }
         public virtual bool UseSchemaFolders { get; set; }
         public virtual bool UseAsyncCalls { get; set; }
+        public virtual bool UseSchemaNamespaces { get; set; }
+
```
Comment (ErikEJ) [inline src/GUI/RevEng.Core.Abstractions/ModuleScaffolderOptions.cs:15]: Nit: remove extra line

File: src/GUI/RevEng.Shared/Cli/Configuration/FileLayout.cs [hunk 5]
```diff
@@ -17,5 +17,9 @@ public class FileLayout
         [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
         [JsonPropertyName("use-schema-folders-preview")]
         public bool UseSchemaFoldersPreview { get; set; }
+
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
+        [JsonPropertyName("use-schema-namespaces-preview")]
```
Comment (ErikEJ) [inline src/GUI/RevEng.Shared/Cli/Configuration/FileLayout.cs:22]: Add this option to the json schema file in the samples folder

File: src/GUI/Shared/Handlers/ReverseEngineer/ReverseEngineerOptions.cs [hunk 6]
```diff
@@ -56,5 +57,6 @@ public class ReverseEngineerOptions
         public bool InstallNuGetPackage { get; set; }
         public bool PreserveCasingWithRegex { get; set; } = true;
         public bool UseDateOnlyTimeOnly { get; set; }
+
```
Comment (ErikEJ) [inline src/GUI/Shared/Handlers/ReverseEngineer/ReverseEngineerOptions.cs:60]: Nit: remove extra line

File: src/GUI/Shared/Models/ModelingOptionsModel.cs [hunk 7]
```diff
@@ -37,6 +37,8 @@ public class ModelingOptionsModel : INotifyPropertyChanged
         private bool useNoObjectFilter;
         private bool useManyToManyEntity;
         private bool useDateOnlyTimeOnly;
+        private bool useSchemaNamespaces;
+
```
Comment (ErikEJ) [inline src/GUI/Shared/Models/ModelingOptionsModel.cs:41]: Remove extra line

**Review by ErikEJ — CHANGES_REQUESTED** [review]

File: src/GUI/RevEng.Shared/Cli/Configuration/FileLayout.cs (hunk 5 above)
Comment (rasputino) [reply src/GUI/RevEng.Shared/Cli/Configuration/FileLayout.cs:22]: Do you mean the json files in the src\GUI\efreveng60\TestFiles folder? I can't find a "samples" folder. Thanks.

Comment (ErikEJ) [reply src/GUI/RevEng.Shared/Cli/Configuration/FileLayout.cs:22]: Sorry, it is this one https://github.com/ErikEJ/EFCorePowerTools/blob/master/samples/efcpt-schema.json

File: samples/efcpt-schema.json [hunk 8]
```diff
@@ -457,13 +457,22 @@
...
+                    ]
                 }
             },
             "examples": [{
                 "output-path": "Models",
                 "output-dbcontext-path": "Context",
                 "split-dbcontext-preview": false,
-                "use-schema-folders-preview": false
+                "use-schema-folders-preview": false,
+                "use-schema-namespaces-preview": false,
```
Comment (ErikEJ) [inline samples/efcpt-schema.json:475]: I think this is invalid json? (Comma at the end)

Comment (rasputino) [reply samples/efcpt-schema.json]: I'm sorry

**Comment by ErikEJ** [comment]
Ping me when you are ready for a second review pass (build is currently failing)

**Comment by rasputino** [comment]
@ErikEJ ping

**Pull request merged**